diff --git a/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/TypeVariableName.kt b/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/TypeVariableName.kt
index 3313b2c..4d19973 100644
--- a/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/TypeVariableName.kt
+++ b/kotlinpoet/src/commonMain/kotlin/com/squareup/kotlinpoet/TypeVariableName.kt
@@ -71,25 +71,36 @@

   override fun emit(out: CodeWriter) = out.emit(name)

+  // Hack for https://github.com/square/kotlinpoet/issues/1737
+  private var eqDepth = 0
+
   override fun equals(other: Any?): Boolean {
-    if (this === other) return true
-    if (javaClass != other?.javaClass) return false
-    if (!super.equals(other)) return false
+    try {
+      // Hack for https://github.com/square/kotlinpoet/issues/1737
+      if (eqDepth++ > 5) return true

-    other as TypeVariableName
+      if (this === other) return true
+      if (javaClass != other?.javaClass) return false
+      if (!super.equals(other)) return false

-    if (name != other.name) return false
-    if (bounds != other.bounds) return false
-    if (variance != other.variance) return false
-    if (isReified != other.isReified) return false
+      other as TypeVariableName

-    return true
+      if (name != other.name) return false
+      if (bounds != other.bounds) return false
+      if (variance != other.variance) return false
+      if (isReified != other.isReified) return false
+
+      return true
+    } finally {
+      eqDepth--
+    }
   }

   override fun hashCode(): Int {
     var result = super.hashCode()
     result = 31 * result + name.hashCode()
-    result = 31 * result + bounds.hashCode()
+    // Hack for https://github.com/square/kotlinpoet/issues/1737
+    // result = 31 * result + bounds.hashCode()
     result = 31 * result + (variance?.hashCode() ?: 0)
     result = 31 * result + isReified.hashCode()
     return result
